src/python-bindings/classad2/exprtree.cpp

```diff
@@ -1 +1,155 @@
+
+    expr->SetParentScope(originalScope);
+    return rv;
+}
+
+
+bool
+evaluate( classad::ExprTree * expr, classad::ClassAd * scope,
+  classad::ClassAd * target, classad::Value & value
+) {
+    if( scope != NULL ) {
+        return EvaluateLooseExpr( expr, scope, target, value );
+    } else if( expr->GetParentScope() ) {
+        return expr->Evaluate( value );
+    } else {
+        classad::EvalState state;
+        return expr->Evaluate( state, value );
+    }
+}
+
+
+
+static PyObject *
+_exprtree_eval( PyObject *, PyObject * args ) {
+    // _exprtree_eval( self._handle, scope )
+
+    PyObject_Handle * self = NULL;
+    PyObject_Handle * scope = NULL;
+    PyObject_Handle * target = NULL;
+    if(! PyArg_ParseTuple( args, "OOO", (PyObject **)& self, (PyObject **)& scope, (PyObject **)& target)) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+
+    auto e = (classad::ExprTree *)self->t;
+    classad::ClassAd * s = NULL;
+    if( scope != NULL && scope != (PyObject_Handle *)Py_None ) {
+        s = (classad::ClassAd *)scope->t;
+    }
+
+    classad::ClassAd * t = NULL;
+    if( target != NULL && target != (PyObject_Handle *)Py_None ) {
+        t = (classad::ClassAd *)target->t;
+    }
+
+
+    classad::Value v;
+    if(! evaluate( e, s, t, v )) {
+        // In version 1, this was ClassAdEvaluationError.
+        PyErr_SetString(PyExc_RuntimeError, "failed to evaluate expression");
+        return NULL;
+    }
+
+    return convert_classad_value_to_python(v);
+}
+
+
+static PyObject *
+_exprtree_simplify( PyObject *, PyObject * args ) {
+    // _exprtree_eval( self._handle, scope )
+
+    PyObject_Handle * self = NULL;
+    PyObject_Handle * scope = NULL;
+    PyObject_Handle * target = NULL;
+    if(! PyArg_ParseTuple( args, "OOO", (PyObject **)& self, (PyObject **)& scope, (PyObject **)& target)) {
+        // PyArg_ParseTuple() has already set an exception for us.
+        return NULL;
+    }
+
+
+    auto e = (classad::ExprTree *)self->t;
+
+    classad::ClassAd * s = NULL;
+    if( scope != NULL && scope != (PyObject_Handle *)Py_None ) {
+        s = (classad::ClassAd *)scope->t;
+    }
+
+    classad::ClassAd * t = NULL;
+    if( target != NULL && target != (PyObject_Handle *)Py_None ) {
+        t = (classad::ClassAd *)target->t;
+    }
+
+
+    // This construction is confusing, but MakeUndefined() returns a new
+    // Literal and overwrites `v` with the address of a member of the return;
+    // we thus don't have to free/delete `v`.
+    //
+    // Not sure why this has to be so hard, but let's not mess with it.
+    classad::Value * v = NULL;
+    classad::Literal * literal = classad::Literal::MakeUndefined(v);
+    if(! evaluate( e, s, t, *v )) {
+        // In version 1, this was ClassAdEvaluationError.
+        PyErr_SetString(PyExc_RuntimeError, "failed to evaluate expression");
+        return NULL;
+    }
+
+    PyObject * rv = py_new_classad_exprtree(literal);
```

Comment:
This is bad if the result of the evaluation is a list or ad. Then, you end up with a Literal that contains a list or ad instead of a literal value. I haven't looked at how this function is used yet, so I don't know would should happen.

Reply:
The most correct thing to do is to use a local Value for the evaluate(). If the result is a classed or list, return a copy. Otherwise, return a Literal constructed from the Value. Since this also appears in v1, I'm ok pushing this off to HTCONDOR-2163.